### PR TITLE
[UR][Loader] Fix handling of native handles

### PR DESCRIPTION
**Pull request by alexbatashev**
Native handles are created by adapters and thus are inheritently backend-specific. Loader can not assume anything about these handles, as even nullptr may be a valid value for such a handle.

This patch changes two things about native handles:
1) Native handles are no longer wrapped in UR objects
2) Dispatch table is extracted from any other argument of the API function

The above is true for all interop APIs except for urPlatformCreateWithNativeHandle, which needs a spec change: https://github.com/oneapi-src/unified-runtime/issues/1068

### CODE REVIEW

File: source/loader/ur_ldrddi.cpp [hunk 1]
```diff
@@ -2601,14 +2540,6 @@ __urdlllocal ur_result_t UR_APICALL urProgramGetNativeHandle(
         return result;
     }
 
-    try {
-        // convert platform handle to loader handle
-        *phNativeProgram = reinterpret_cast<ur_native_handle_t>(
```
Comment (jandres742) [inline source/loader/ur_ldrddi.cpp:2606]: @bmyates : why is that we need this kind of logic in the L0 loader, of translating driver handles to L0 handles? Do you see any risks on removing this code in the UR loader?

Comment (pbalcer) [reply source/loader/ur_ldrddi.cpp:2606]: The loader handles are there to find the appropriate adapter-level function to call. The alternative is to either make all functions accept an adapter handle as one of the arguments or create a way to map normal handles to adapter handles.

This mechanism is not used when there's only one adapter, see https://github.com/oneapi-src/unified-runtime/issues/355.

Comment (alexbatashev) [reply source/loader/ur_ldrddi.cpp:2606]: > The alternative is to either make all functions accept an adapter handle as one of the arguments or create a way to map normal handles to adapter handles.

There's no need to. Most interop calls accept a context, which has the right adapter, and this patch exactly changes the behavior to find the adapter in the context, rather the native handle. The only notable exception is platform: https://github.com/oneapi-src/unified-runtime/issues/1068. That one certainly needs an adapter as an argument.

**Review by pbalcer — APPROVED** [review]
Please also modify [`get_adapter_handles`](https://github.com/oneapi-src/unified-runtime/blob/adapters/scripts/templates/helper.py#L579) so that it doesn't return native handles. This will remove the now unused object factories.

File: source/loader/ur_ldrddi.cpp (hunk 1 above)
Comment (pbalcer) [reply source/loader/ur_ldrddi.cpp:2606]: > There's no need to.

I know, I agree. I was replying to Jamie's question.

**Comment by alexbatashev** [comment]
> Please also modify [`get_adapter_handles`](https://github.com/oneapi-src/unified-runtime/blob/adapters/scripts/templates/helper.py#L579) so that it doesn't return native handles. This will remove the now unused object factories.

I did try to change it to:

```python
def get_adapter_handles(specs):
    objs = []
    for s in specs:
        for obj in s['objects']:
            if obj_traits.is_handle(obj) and not (obj_traits.is_loader_only(obj) or 'native' in obj['name']):
                objs.append(obj)

    return objs
```

But the compilation would fail with the following error:

```
def get_adapter_handles(specs):
    objs = []
    for s in specs:
        for obj in s['objects']:
            if obj_traits.is_handle(obj) and not (obj_traits.is_loader_only(obj) or 'native' in obj['name']):
                objs.append(obj)

    return objs
```

Am I doing something wrong?

File: source/loader/ur_ldrddi.cpp (hunk 1 above)
Comment (jandres742) [reply source/loader/ur_ldrddi.cpp:2606]: thanks @pbalcer , @alexbatashev . Ah, this is only for native handles. That's ok then.

+1

**Comment by pbalcer** [comment]
> Am I doing something wrong?

No, I think it's all correct. The failure is because platform native handles are special-cased for now. This particular change will have to wait until the spec is fixed.

**Comment by alexbatashev** [comment]
> > Am I doing something wrong?
> 
> No, I think it's all correct. The failure is because platform native handles are special-cased for now. This particular change will have to wait until the spec is fixed.

Well, I guess, that is the reason why we can't remove `ur_native_handle_t` factory for now. I don't see a way to leave it only for urProgramCreateWithNative.

**Pull request merged**